### ISSUE TITLE
turning off kv cache?

### ISSUE THREAD

**Issue by blenderman94** [LostRuins/koboldcpp#1848]
greetings friend. im wondering about how i may can turn off kv caching altogether? i know that it should be making things faster and everythin. but it mostly makes clutter a lot. and wasted a tons of memory. reasoning is that because if it you can put less layers on gpu but because you put less layers there it will be slow anyways. just with more ram wasted.

**Comment by MathapMathonwy** [comment]
No. The KV cache is critical to how the LLM actually works. Your options are to either put KV cache on system ram (this is the "Low VRAM" option.), or use flash attention and quantize the KV-cache which will reduce its size, but can significantly impact accuracy. 

Just guessing, but you'd probably benefit from trying to use a smaller model. Vast progress has been made since Vicuna. Mistral-Nemo at 12B parameters (and while fairly old itself) is still widely used, and will be *much* more performant than a 30B, and also natively supports higher context.  Or you could look at Qwen3-14B or smaller.

**Comment by blenderman94** [comment]
im usualy like to do things what my pc not meant to as you may guessed. but i also see no option in the kobold cpp gui. should i start with a launch parameter instead ? i personaly tryed qwen and many other from 8-30 b and what was common in all of them that they was dumb. (or maybe i having high expectations) not i actualy trying something larger. name comand r 35b

**Comment by LostRuins** [comment]
In the GUI launcher, you can turn it off in the Hardware tab, look for "No KV offload" which will avoid offloading kv

**Comment by blenderman94** [comment]
then pls gave me a generic template command to. low vram ,model load ,tts load

**Comment by LostRuins** [comment]
These are the basic flags that should work when launching from cli

-nvko
--ttsmodel (your tts model.gguf)
--model (your model.gguf)
--usecuda
--gpulayer 99

**Comment by HumbleDeer** [comment]
> -nvko

Are you sure this is the one? In my head, shortening `No KV Offload` results in a different order of letters.. 🙈

**Comment by HumbleDeer** [comment]
To elaborate on the core question as well:

OP:

> how i may can turn off kv caching altogether?

Reply:

> No. The KV cache is critical to how the LLM actually works.

If we assume that at first this was actually to be taken literally and not as a product of a language barrier...
The reason why you can't disable the KV cache, is because the KV part stands for Key-Value, and those keys with values are integral to the LLM and inference process. In the past, indeed, it wasn't a thing. However, since it has become so widely recognized there's no reason not to use this specific addition to the internal workings, it is nowadays considered a "core" part of the internal workings.

So what does it do?
This requires a brief mention of something else: Attention. It's a processed pipeline of sorts, where your input text is combined with all the mathematical relationships of words and how they relate to each other -- all the complicated stuff, but then placed in a format that is utmost-efficient for the LLM's internal mechanism to use for "remembering". It is integral to its ability to generate a response, because without it, there's nothing to generate a response to. 

If you've ever heard of the term Context in this sphere, then you'll be familiar with how you personally interact with it. The Context mechanism is a broader category of underlying tools that all together get called the context machine or something similar. Key-Value pairs saved in memory are a large part of what the concept of "Context" is made up of, although not the only thing. Yet, everything that the Context "does" for the start to end text generation procedure ends up going or out or through the K-V cache. Because _Everything Affects Everything!_. It's not easy to simplify, so I'm hoping this is a little clearer without having you learn how transformers with attention blocks work. 

(Tangent extras:)
! **Oh! blocks**: that's a term we use for one of the building blocks that go into a "layer". A block is a part of the machine that has rows and columns, and many rows and many columns together form a layer. Then you stack layers on top of each other so that they can be done in sequence. 

> Just guessing, but you'd probably benefit from trying to use a smaller model. Vast progress has been made since Vicuna. Mistral-Nemo at 12B parameters (and while fairly old itself) is still widely used, and will be _much_ more performant than a 30B, and also natively supports higher context. Or you could look at Qwen3-14B or smaller.

It's in fact much wiser to pick a smaller model at a lower quantization level (higher number) because that means you're not "blindly" throwing out detail as much as quantization does, in effect. In reality, it's more complicated, but that's the simple idea. If you have the option of using a 7/8B model and fit it all in your VRAM at Q4 or Q5, then that's vastly preferred over a 14B or 20B model at Q2 just so it would fit in memory. 

The more you quantize (remove precision of) data, the more you're blindly handicapping its ability to work. The level of mathematical confusion is a lot higher when you try and run a large model at a tiny quantized size. If you start out with a model that's trained at a smaller size, you don't have nearly as much handicapping of its ability to put 2 and 2 together internally in the most natural flowing pattern, which means it might very well perform nicer than a larger model that's been handicapped deeper.

**Comment by blenderman94** [comment]
I dont know I coded a Few small llms in tensorflow as well in pytorch. That I nevet ever needed a kv cache (it should just to speed up compute) it's maybe a backend only thing? That couple gigs of ram could make a lot of wasted ram even on ddr5 for possibly minor return
rns and bloat off course its for cpu guys gpu is nice but it's generally a nuisance (expect on I gpu with a lot of shared ram) an ai model is just a bunch of tensors. And ops between tensors te thing might be not truly needed. But I won't say it without some usefulness

**Comment by wbruna** [comment]
> Are you sure this is the one? In my head, shortening `No KV Offload` results in a different order of letters.. 🙈

It's "-nkvo" actually 🙂
https://github.com/LostRuins/koboldcpp/blob/9720aa6224ae26d40cd3703ac4cf4af7d2a94450/koboldcpp.py#L7911

**Comment by HumbleDeer** [comment]
> It's "-nkvo" actually 🙂

I already figured as much 😅

**Comment by LostRuins** [comment]
sorry my finger slipped 💀

**Comment by wethecom** [comment]
coming from a third party turning of key values caching has tested to make  the system run faster and not lose quality

https://www.youtube.com/watch?v=TYgCRPCAFhE

**Comment by blenderman94** [comment]
well like what i was saying from the first time i opened this thing. kv cache is while seemingly good actually stupid. think for a moment like if you trying to a job and you keep bothering about what happened month ago. also from a programing perspective its kinda stupid too. mostly large caches are. good job wethecom! the idea as i was undertood is to save some resources on attention. but for that you want to compute the attention smarter in a tiled fashion instead. then using a big bag what the kv is mostly becasue sometimes you want to think it over. but most importantly you hold hold back the gpu un necesarily.
